pkg/common/cns-lib/volume/manager.go

```diff
@@ -515,6 +562,79 @@ func (m *defaultManager) createVolumeWithImprovedIdempotency(ctx context.Context
 	return resp, faultType, err
 }
 
+func (m *defaultManager) waitOnTask(csiOpContext context.Context,
+	taskMoRef vim25types.ManagedObjectReference) (*vim25types.TaskInfo, error) {
+	log := logger.GetLogger(csiOpContext)
+	if m.listViewIf == nil {
+		err := m.initListView()
+		if err != nil {
+			return nil, err
+		}
+	}
+	ch := make(chan Result)
+	err := m.listViewIf.AddTask(csiOpContext, taskMoRef, ch)
+	if err != nil {
+		return nil, logger.LogNewErrorf(log, "failed to add task to list view. err: %v", err)
+	}
+
+	// deferring removal of task after response from CNS
+	// called only after successful addition to listView, so we don't need any check for the removal
+	defer func() {
+		err := m.listViewIf.RemoveTask(csiOpContext, taskMoRef)
+		if err != nil {
+			log.Errorf("failed to remove task from list view. err: %v", err)
+			err = m.listViewIf.MarkTaskForDeletion(csiOpContext, taskMoRef)
+			if err != nil {
+				log.Errorf("failed to mark task for deletion. err: %v", err)
+			}
+		}
+	}()
+
+	return waitForResultOrTimeout(csiOpContext, ch)
+}
+
+// waitForResultOrTimeout uses the context provided by the sidecars when CSI driver operations are called.
+// This context has a timeout associated with it (see manifests for more details).
+// Once this caller timeout is over, we want to return an error back to the caller
+func waitForResultOrTimeout(csiOpContext context.Context, ch chan Result) (*vim25types.TaskInfo, error) {
+	var taskInfo *vim25types.TaskInfo
+	var err error
+	select {
+	case <-csiOpContext.Done():
+		err = errors.New("caller timed out before response from CNS")
+		taskInfo = nil
+	case result := <-ch:
+		err = result.Err
+		taskInfo = result.TaskInfo
+	}
+	return taskInfo, err
+}
+
+func (m *defaultManager) initListView() error {
+	ctx := logger.NewContextWithLogger(context.Background())
+	log := logger.GetLogger(ctx)
+	log.Debugf("Initializing new listView object for vc: %+v", m.virtualCenter)
+	if m.virtualCenter.Client == nil {
+		log.Debugf("vimClient is nil. calling vc.connect()")
+		err := m.virtualCenter.Connect(ctx)
+		if err != nil {
+			return logger.LogNewErrorf(log, "failed to connect to vCenter. err: %v", err)
+		}
+	}
+
+	govmomiClient, err := m.virtualCenter.NewClient(ctx)
+	if err != nil {
+		return logger.LogNewErrorf(log, "failed to create a separate govmomi client for listView. error: %+v", err)
+	}
+	govmomiClient.Timeout = noTimeout
```

Comment:
what is the default if not set? Do we require to set this to 0

Reply:
Setting the http.Client.Timeout to 0 enables us to WaitForUpdates() without getting a context deadline exceeded error